extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java

```diff
@@ -0,0 +1,113 @@
+package io.quarkus.arc.runtime;
+
+import java.lang.annotation.Annotation;
+import java.util.Set;
+
+import jakarta.annotation.Priority;
+import jakarta.enterprise.inject.spi.Annotated;
+import jakarta.enterprise.inject.spi.AnnotatedConstructor;
+import jakarta.enterprise.inject.spi.AnnotatedField;
+import jakarta.enterprise.inject.spi.AnnotatedParameter;
+import jakarta.enterprise.inject.spi.InjectionPoint;
+import jakarta.inject.Inject;
+import jakarta.interceptor.AroundInvoke;
+import jakarta.interceptor.Interceptor;
+import jakarta.interceptor.InvocationContext;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.config.NativeBuildTime;
+import io.quarkus.arc.impl.InjectionPointProvider;
+import io.quarkus.runtime.ImageMode;
+
+/**
+ * The goal of this interceptor is to verify the current ImageMode when a dependent config property is being injected.
+ */
+@Priority(jakarta.interceptor.Interceptor.Priority.PLATFORM_BEFORE)
+@Interceptor
+@NativeBuildConfigCheck
+public class NativeBuildConfigCheckInterceptor {
+
+    private static final Logger LOG = Logger.getLogger(NativeBuildConfigCheckInterceptor.class);
+
+    @Inject
+    NativeBuildConfigContext nativeBuildConfigContext;
+
+    @AroundInvoke
+    Object aroundInvoke(InvocationContext context) throws Exception {
+        verifyCurrentImageMode(nativeBuildConfigContext.getBuildAndRunTimeFixed());
+        return context.proceed();
+    }
+
+    static void verifyCurrentImageMode(Set<String> buildAndRunTimeFixed) {
+        InjectionPoint injectionPoint = InjectionPointProvider.get();
+        if (injectionPoint != null) {
+            // Skip injection points annotated with NativeBuildTime
+            Annotated annotated = injectionPoint.getAnnotated();
+            if (annotated != null && annotated.isAnnotationPresent(NativeBuildTime.class)) {
+                return;
+            }
+            // Skip BUILD_AND_RUN_TIME_FIXED properties
+            if (!buildAndRunTimeFixed.isEmpty()) {
+                String propertyName = null;
+                for (Annotation qualifier : injectionPoint.getQualifiers()) {
+                    if (qualifier instanceof ConfigProperty) {
+                        propertyName = ((ConfigProperty) qualifier).name();
+                    }
+                }
+                if (propertyName != null && buildAndRunTimeFixed.contains(propertyName)) {
+                    return;
+                }
+            }
+        }
+        if (ImageMode.current() == ImageMode.NATIVE_BUILD) {
```

Comment:
Would it make sense to put this check at the very top, to avoid unnecessary computation at (native) runtime?

Or maybe you could add a GraalVM substitution to make this method a no-op at native runtime?

Reply:
I will put it at the very top. Although I don't think it will be a measurable improvement ;-)

> Or maybe you could add a GraalVM substitution to make this method a no-op at native runtime?

I don't think it's worth the effort.